pandas/_libs/internals.pyx

```diff
@@ -890,17 +890,25 @@ cdef class BlockValuesRefs:
     """
     cdef:
         public list referenced_blocks
+        public int clear_counter
 
     def __cinit__(self, blk: Block | None = None) -> None:
         if blk is not None:
             self.referenced_blocks = [weakref.ref(blk)]
         else:
             self.referenced_blocks = []
+        self.clear_counter = 500  # set reasonably high
 
-    def _clear_dead_references(self) -> None:
-        self.referenced_blocks = [
-            ref for ref in self.referenced_blocks if ref() is not None
-        ]
+    def _clear_dead_references(self, force=False) -> None:
+        if force or len(self.referenced_blocks) > self.clear_counter:
+            self.referenced_blocks = [
+                ref for ref in self.referenced_blocks if ref() is not None
+            ]
+            nr_of_refs = len(self.referenced_blocks)
+            if nr_of_refs < self.clear_counter // 2:
+                self.clear_counter = self.clear_counter // 2
+            elif nr_of_refs > self.clear_counter:
+                self.clear_counter = min(self.clear_counter * 2, nr_of_refs)
```

Comment:
Could also just do x2 instead of the `min(..)` calculation? 

I am wondering that if you repeatedly add a reference (for an object that doesn't go out of scope), doesn't that end up increasing the counter only with +1 every time? For example, you have 501 refs, at that moment you clear the refs, but `nr_of_refs` is still 501 one after doing that, and then here we set the new threshold to `min(500 * 2, 501)`, i.e. 501? 
I must be missing something, because otherwise I don't understand how this fixes the perf issue of the `list(df.items())` example.

Reply:
Yeah you are correct, already added a max here because I came to the same conclusion